### PR TITLE
DCOS-10470: [4/10] Introducing pluggable DSL filters

### PR DESCRIPTION
**Pull request by wavesoft**
---
⚠️  Depends on #1493

---

[Diff between 1493 & This](https://github.com/dcos/dcos-ui/compare/ic/feat/DCOS-10470-new-dsl-components-3...ic/feat/DCOS-10470-new-dsl-components-4?expand=1)

This PR introduces pluggable `DSLFilter` objects and the `DSLFilters` container that is used by the DSL parser to compose the final filter function.

### CODE REVIEW

**Comment by mlunoe** [comment]
**mesosphere-ci** Retest this please

**Comment by mlunoe** [comment]
Hmm... seems like the `./scripts/validate-tests` fails for some reason

File: src/js/structs/DSLFilter.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  /* eslint-disable no-unused-vars */
+
+  /**
+   * Returns true if the filter can be applied to the given operand
+   *
+   * @param {DSLFilterTypes} filterType - The filter type (ex. fuzzy or attrib)
+   * @param {Object} filterArguments - The filter arguments
+   * @returns {Boolean} True if the filter can be applied to the given operand
+   */
+  filterCanHandle(filterType, filterArguments) {
```
Comment (mlunoe) [inline src/js/structs/DSLFilter.js]: Why are these functions empty?
Is this file implemented fully?

Comment (wavesoft) [inline src/js/structs/DSLFilter.js]: This is an abstract method that may or may not be implemented by the DSL filter.
The full signature and the comment block in the function is for documentation purposes.

File: src/js/structs/DSLFilters.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/**
+ * This class maintains a dictionary with all the currently available
+ */
+class DSLFilters {
```
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: Could we call this `DSLFilterList` as we do other structs. This makes it easier to differentiate between the two

File: src/js/structs/DSLFilters.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    this.filters = filters;
+  }
+
+  /**
+   * Plug a filter to the list of filters
+   *
+   * @param {DSLFilter} filter - The filter to plug
+   * @returns {DSLFilters} Returns a DSLFilters object for chaining the calls
+   */
+  plug(filter) {
```
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: Should we make this take more filters?
```js
plug(...args) {
  this.filters.push(...args);

  return this;
}

File: src/js/structs/DSLFilters.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    this.filters.push(filter);
+    return this;
```
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: #nit newline before `return`

File: src/js/structs/DSLFilters.js (hunk 3 above)
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: Should we throw an error in development mode, if something that is not a DSLFilter is passed?

File: src/js/utils/DSLParserUtil.js [hunk 5]
```diff
@@ -98,16 +93,33 @@ function filterFunctionFactory(ast) {
    * @returns {List} resultset - A new instance of a List, containing the results
    */
   return function (filters, resultset) {
+    if (!(filters instanceof DSLFilters)) {
```
Comment (mlunoe) [inline src/js/utils/DSLParserUtil.js]: Could we get some tests for this function?

File: src/js/utils/DSLParserUtil.js [hunk 6]
```diff
@@ -60,14 +62,8 @@ function combineFunctionFactory(ast, leftFilterFn, rightFilterFn) {
...
-
-        // TODO: Merge the results of `leftFilterFn(filters, resultset)` and
-        //       `rightFilterFn(filters, resultset)`.
-
-        // TODO: Implement a `List.merge` function that merges two lists,
-        //       discarding duplicate values.
-
-        return resultset;
+        let intermediateResultset = leftFilterFn(filters, resultset);
+        return intermediateResultset.combine(rightFilterFn(filters, resultset));
```
Comment (mlunoe) [inline src/js/utils/DSLParserUtil.js]: #nit newline before `return`

File: src/js/utils/DSLParserUtil.js [hunk 7]
```diff
@@ -60,14 +62,8 @@ function combineFunctionFactory(ast, leftFilterFn, rightFilterFn) {
...
         // We are interested in the union of the results of the two filters
-
-        // TODO: Merge the results of `leftFilterFn(filters, resultset)` and
-        //       `rightFilterFn(filters, resultset)`.
-
-        // TODO: Implement a `List.merge` function that merges two lists,
-        //       discarding duplicate values.
-
-        return resultset;
+        let intermediateResultset = leftFilterFn(filters, resultset);
```
Comment (mlunoe) [inline src/js/utils/DSLParserUtil.js]: Could we get some tests for this function?

**Comment by mlunoe** [comment]
Rebased to make build pass again after this addition: https://github.com/dcos/dcos-ui/pull/1494/commits/5631732b33992876880ee59a55f149b8e8e2eff8

File: src/js/structs/DSLFilters.js (hunk 3 above)
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: done!

File: src/js/structs/DSLFilters.js (hunk 4 above)
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: Done!

File: src/js/utils/DSLParserUtil.js (hunk 7 above)
Comment (mlunoe) [inline src/js/utils/DSLParserUtil.js]: Done!

File: src/js/utils/DSLParserUtil.js (hunk 6 above)
Comment (mlunoe) [inline src/js/utils/DSLParserUtil.js]: Done!

File: src/js/structs/DSLFilters.js (hunk 2 above)
Comment (mlunoe) [inline src/js/structs/DSLFilters.js]: Done

**Comment by wavesoft** [comment]
retest this please

**Comment by wavesoft** [comment]
Addressed 👍

**Pull request merged**